Add Continuity Check Script to celo-migrate

alecps opened the pull request:
Adds a standalone command for checking db continuity without performing a migration.

Unit tests are added for the CheckContinuity function

The continuity script can be run in "fail-fast" mode to quickly check whether the db is corrupted, or run normally to print out every gap detected in the data.

Drive by changes: 
- Adds fixes for a couple minor indexing errors
- Typos
- Fixes a couple error checks (fixing these types of error checks was part of the audit feedback, but I found a couple that were missed) 

Fixes https://github.com/celo-org/celo-blockchain-planning/issues/832

Note: This PR originally added the continuity check logic to the actual migration code path (in addition to having a standalone script), but in an effort to reduce unnecessary changes to the migration logic those changes were reverted and moved to https://github.com/celo-org/optimism/pull/297. This means that the migration script itself will not detect any gaps in the data being migrated, so the standalone continuity script should be run on the source db before running the migration script.

palango commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -144,6 +144,9 @@ func readAncientBlocks(ctx context.Context, freezer *rawdb.Freezer, startBlock,
 	// Transform blocks from the in channel and send them to the out channel
 	defer close(out)
+
+	prevBlockNumber := uint64(0)
```
This relies on the assumption that the incoming block ranges are ordered, which is not obvious. Maybe add a comment?

palango commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -152,6 +155,11 @@ func transformBlocks(ctx context.Context, in <-chan RLPBlockRange, out chan<- RL
 				blockNumber := blockRange.start + uint64(i)
 
+				if blockNumber != 0 && blockNumber != prevBlockNumber+1 {
+					return fmt.Errorf("Gap found between ancient blocks numbered %d and %d. Please delete the target directory and repeat the migration with an uncorrupted source directory.", prevBlockNumber, blockNumber)
```
```suggestion
					return fmt.Errorf("gap found between ancient blocks numbered %d and %d. Please delete the target directory and repeat the migration with an uncorrupted source directory.", prevBlockNumber, blockNumber)
```

palango commented on op-chain-ops/cmd/celo-migrate/non-ancients.go:
```diff
@@ -75,11 +75,18 @@ func migrateNonAncientsDb(newDB ethdb.Database, lastBlock, numAncients, batchSiz
 		log.Info("Processing Block Range", "process", "non-ancients", "from", i, "to(inclusve)", i+batchSize-1, "count", len(numbersHash))
 		for _, numberHash := range numbersHash {
+			if numberHash.Number != 0 && numberHash.Number != prevBlockNumber+1 {
+				return 0, fmt.Errorf("Gap found between non-ancient blocks numbered %d and %d. Please delete the target directory and repeat the migration with an uncorrupted source directory.", prevBlockNumber, numberHash.Number)
```
```suggestion
				return 0, fmt.Errorf("gap found between non-ancient blocks numbered %d and %d. Please delete the target directory and repeat the migration with an uncorrupted source directory.", prevBlockNumber, numberHash.Number)
```

palango reviewed (commented):
Looks good, but I wonder if we should have a separate command to check a datadir before the migration, independent of the pre-migration. What do you think?

alecps commented:
@palango Interesting idea. I need to think about that a bit more. My first reaction is that it would add an extra step + time to the migration process. I'm also trying to work out if it's even possible to have gaps in ancient blocks, as the ancient db is append only. Having a separate script could be a quick way to bring extra piece of mind. Will give this some more thought tomorrow

palango commented:
> @palango Interesting idea. I need to think about that a bit more. My first reaction is that it would add an extra step + time to the migration process. I'm also trying to work out if it's even possible to have gaps in ancient blocks, as the ancient db is append only. Having a separate script could be a quick way to bring extra piece of mind. Will give this some more thought tomorrow

Not a requirement, but it might be too late for a full resync if people run the pre-migration shortly before the schedules block.

piersy commented:
My feeling is that a separate command for this is unnecessary, if we want people to check this earlier then we should just message them to run the pre-migration earlier.

piersy commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -152,6 +155,12 @@ func transformBlocks(ctx context.Context, in <-chan RLPBlockRange, out chan<- RL
 			for i := range blockRange.hashes {
 				blockNumber := blockRange.start + uint64(i)
 
+				if blockNumber != 0 && blockNumber != prevBlockNumber+1 {
+					return fmt.Errorf("gap found between ancient blocks numbered %d and %d. Please delete the target directory and repeat the migration with an uncorrupted source directory.", prevBlockNumber, blockNumber)
+				}
```
I don't think this is going to work, If you are doing a second pre-migrate and therefore `blockNumber` is not 0 then this condition will be triggered.

alecps replied:
Good catch!

piersy replied:
If `startBlock` can be 0, which I think it can for the first block then `prevBlockNumber` will underflow. But actually I think that is ok, because when the check will cause it to overflow back by adding 1.

alecps commented:
I've observed some strange behavior that I want to look into more. It seems there may be an earlier gap that this branch does not detect that is preventing blocks from freezing beyond block `5,002,000`. This branch only throws on a later gap that starts at block `5835918`. It's possible this is a partial gap, where only receipt data is missing. Will re-open this for review when I've gotten to the bottom of what's going on

piersy commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -152,6 +155,12 @@ func transformBlocks(ctx context.Context, in <-chan RLPBlockRange, out chan<- RL
 			for i := range blockRange.hashes {
 				blockNumber := blockRange.start + uint64(i)
 
+				if blockNumber != 0 && blockNumber != prevBlockNumber+1 {
```
This check for zero is redundant now
```suggestion
				if blockNumber != prevBlockNumber+1 {
```

piersy commented on op-chain-ops/cmd/celo-migrate/non-ancients.go:
```diff
@@ -75,11 +75,18 @@ func migrateNonAncientsDb(newDB ethdb.Database, lastBlock, numAncients, batchSiz
 
 		log.Info("Processing Block Range", "process", "non-ancients", "from", i, "to(inclusve)", i+batchSize-1, "count", len(numbersHash))
 		for _, numberHash := range numbersHash {
+			if numberHash.Number != 0 && numberHash.Number != prevBlockNumber+1 {
```
This check for zero is also redundant
```suggestion
			if numberHash.Number != prevBlockNumber+1 {
```

palango commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -141,11 +141,11 @@ func readAncientBlocks(ctx context.Context, freezer *rawdb.Freezer, startBlock,
 }
 
-func transformBlocks(ctx context.Context, in <-chan RLPBlockRange, out chan<- RLPBlockRange) error {
+func transformBlocks(ctx context.Context, in <-chan RLPBlockRange, out chan<- RLPBlockRange, startBlock uint64) error {
```
More a theoretical question, what happens if there are no ancients in the db when this is run. Wouldn't this underflow?

alecps replied:
Yes, but it's actually okay because it overflows again in the actual check. I added a comment clarifying this

piersy replied:
@palango I don't think the script works if there are no ancients, there are a number of places.

Checked with no-ancients
```
./bin/celo-migrate full --old-db ../../chaindatas/alfajores_small/celo/chaindata --new-db alfajores_small_migrated --deploy-config ../../celo-l2-node-docker-compose/envs/alfajores/migration-config/config.json  --l1-deployments ../../celo-l2-node-docker-compose/envs/alfajores/migration-config/deployment-l1.json --l2-allocs ../../celo-l2-node-docker-compose/envs/alfajores/migration-config/l2-allocs.json --l1-rpc https://ethereum-holesky-rpc.publicnode.com --outfile.rollup-config out-rollup.json --outfile.genesis out-genesis.json --migration-block-number 71458 --migration-block-time `date +%s`
INFO [12-18|15:52:50.823] Full Migration Started                   oldDBPath=../../chaindatas/alfajores_small/celo/chaindata newDBPath=alfajores_small_migrated
INFO [12-18|15:52:50.823] Allocated cache and file handles         database=../../chaindatas/alfajores_small/celo/chaindata cache=1024.00MiB handles=60 readonly=true
INFO [12-18|15:52:50.827] Using LevelDB as the backing database
INFO [12-18|15:52:50.830] Source db is synced to correct height    head=71457 migrationBlock=71458
INFO [12-18|15:52:50.830] Pre-Migration Started                    oldDBPath=../../chaindatas/alfajores_small/celo/chaindata newDBPath=alfajores_small_migrated batchSize=50000 memoryLimit=7500
INFO [12-18|15:52:50.832] Copying files from old database (excluding ancients) process=non-ancients
INFO [12-18|15:52:50.841] Opened ancient database                  database=../../chaindatas/alfajores_small/celo/chaindata/ancient readonly=false
INFO [12-18|15:52:50.882] Running rsync command                    command="/usr/bin/rsync -v -a --exclude=ancient --checksum --delete --progress ../../chaindatas/alfajores_small/celo/chaindata/ alfajores_small_migrated"
INFO [12-18|15:52:50.885] Opened ancient database                  database=alfajores_small_migrated/ancient readonly=false
INFO [12-18|15:52:50.885] Ancient Block Migration Skipped          process=ancients     ancientsInOldDB=0 ancientsInNewDB=0
building file list ...
INFO [12-18|15:52:50.902] TIMER                                    process=ancients     duration=69.081958ms
INFO [12-18|15:52:50.902] Using leveldb as the backing database
INFO [12-18|15:52:50.902] Allocated cache and file handles         database=../../chaindatas/alfajores_small/celo/chaindata cache=1024.00MiB handles=60 readonly=true
INFO [12-18|15:52:50.904] Using LevelDB as the backing database
INFO [12-18|15:52:50.904] Found legacy ancient chain path          location=../../chaindatas/alfajores_small/celo/chaindata/ancient
INFO [12-18|15:52:50.904] Opened ancient database                  database=../../chaindatas/alfajores_small/celo/chaindata/ancient readonly=true
INFO [12-18|15:52:50.907] TIMER                                    process=getStrayAncientBlocks duration=5.420459ms
panic: runtime error: makeslice: cap out of range

goroutine 4 [running]:
github.com/ethereum/go-ethereum/core/rawdb.ReadAllHashesInRange({0x106006378, 0x140003b28d0}, 0x104c01f01?, 0xffffffffffffffff)
	/Users/pierspowlesland/go/pkg/mod/github.com/celo-org/op-geth@v1.101408.1-0.20241209113131-e6499495c176/core/rawdb/accessors_chain.go:93 +0x98
main.getStrayAncientBlocks({0x16bbfe3f1, 0x2f})
	/Users/pierspowlesland/projects/optimism/op-chain-ops/cmd/celo-migrate/ancients.go:293 +0x234
main.runPreMigration.func1()
	/Users/pierspowlesland/projects/optimism/op-chain-ops/cmd/celo-migrate/main.go:301 +0xf4
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/pierspowlesland/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
	/Users/pierspowlesland/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:75 +0x98
14 files to consider
./
```

piersy replied:
But I don't think we should be fixing this, we are always going to have ancients

alecps commented:
I added checks for all the necessary data (receipts, headers, total difficulty, etc.) for both ancient and non-ancient blocks. The script before only checked for gaps in headers. I haven't been able to test with gaps in ancients because I haven't been able to get a node to freeze blockls beyond the first gap. It's unclear whether it's possible to actually have a gap in the ancient data.

alecps commented:
@palango After thinking about it more, I do actually think it would be nice to have a command for testing data integrity. Though it isn't strictly necessary, it might be appreciated by anyone who runs into data corruption issues and wants to quickly check whether the database they're using is okay before trying a second time. We could even consider recommending that people run it first as a precaution, but that might not be necessary. I think it might make sense to add in a follow up PR so that this one doesn't get too big. What do you think?

palango commented:
> @palango After thinking about it more, I do actually think it would be nice to have a command for testing data integrity. Though it isn't strictly necessary, it might be appreciated by anyone who runs into data corruption issues and wants to quickly check whether the database they're using is okay before trying a second time. We could even consider recommending that people run it first as a precaution, but that might not be necessary. I think it might make sense to add in a follow up PR so that this one doesn't get too big. What do you think?

Totally agree, this can be added separately. And maybe this doesn't need to be coupled to the migration at all, even though it would make sense to reuse code if possible.

piersy commented on op-chain-ops/cmd/celo-migrate/non-ancients.go:
```diff
@@ -125,3 +142,34 @@ func migrateNonAncientBlock(number uint64, hash common.Hash, newDB ethdb.Databas
 	return nil
 }
+
+// checkOtherDataForNonAncientBlock checks that all the data that is not transformed is succesfully copied for non-ancient blocks.
```
```suggestion
// checkOtherDataForNonAncientBlock checks that all the data that is not transformed is successfully copied for non-ancient blocks.
```

piersy commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -23,6 +24,44 @@ type RLPBlockRange struct {
+		return err
+	}
+
+	// Cbecm that block number in header matches the expected block number
```
```suggestion
	// Check that block number in header matches the expected block number
```

piersy commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -144,15 +183,22 @@ func readAncientBlocks(ctx context.Context, freezer *rawdb.Freezer, startBlock,
 				return fmt.Errorf("failed to read tds from old freezer: %w", err)
 			}
 
+			if err := CheckRLPBlockRangeForGaps(blockRange, count); err != nil {
+				return fmt.Errorf("failed to ensure ancient block range has no gaps: %w", err)
+			}
+
```
I think it would be simpler and safer if you checked for consistency between block ranges here, instead of in  `transformBlocks`.

You could adjust the Check to accept two ranges and check the transition between them. E.g

`CheckRLPBlockRangeForGaps(prevBlockRange, currBlockRange RLPBlockRange, expectedLength uint64)`

If prev is non nil you check that the last block of prev is the parent of the first block of curr.

piersy commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -23,6 +24,44 @@ type RLPBlockRange struct {
+		expectedBlockNumber := blockRange.start + i
+		if header.Number.Uint64() != expectedBlockNumber {
+			return fmt.Errorf("header number mismatch: expected %d, actual %d", expectedBlockNumber, header.Number.Uint64())
+		}
```
I would also check for parent header consistency, so check that the parentHash of block `N` equals the hash of  block `N-1`.

alecps replied:
So I added this check and it appears the parent hashes don't match the previous block hash. I guess this makes sense, as we've changed the format of the header but not the ParentHash field. @piersy Do you think this is something we should transform in the block header or leave as is?

alecps replied:
Nevermind, figured out that this mismatch was due to hashing over the non-transformed header. I fixed the check and added a comment explaining this.

alecps commented on op-chain-ops/cmd/celo-migrate/main.go:
```diff
@@ -449,6 +497,140 @@ func runStateMigration(newDBPath string, opts stateMigrationOptions) error {
+			count = count + 1
+		}
+		g.Go(func() error {
+			select {
```
@piersy before you dig into the rest, would love a quick check on whether I've done everything right here re using cancel(), ctx etc.

piersy commented on op-chain-ops/cmd/celo-migrate/continuity_test.go:
```diff
@@ -0,0 +1,207 @@
+				start:    0,
+				hashes:   [][]byte{[]byte("hash0"), []byte("hash1"), []byte("hash2"), []byte("hash3")},
+				headers:  encodedHeaders,
+				bodies:   [][]byte{[]byte("body0"), []byte("body1"), []byte("body2"), []byte("body3")},
+				receipts: [][]byte{[]byte("receipt0"), []byte("receipt1"), []byte("receipt2"), []byte("receipt3")},
+				tds:      [][]byte{[]byte("td0"), []byte("td1"), []byte("td2"), []byte("td3")},
```
Adding all these fields here is verbose and confusing because it makes it look like the test is checking some of these values when it is not.

I suggest using a function like `makeRange(start int, hashes, encodedHeaders [][]byte)` that ensures that all other fields have the correct length

alecps replied:
Got it. Let me know if what I have now looks better to you. I used a makeRange helper to make the tests more concise, but since many of the tests are specifically checking for differences in the lengths of the various other fields, I'm still passing the slices in to the helper for all the fields (not just  hashes and encodedHeaders). This ended up looking easier to read in my opinion than passing the desired lengths as int parameters. I can't see how the parameters you suggested above "`makeRange(start int, hashes, encodedHeaders [][]byte)`" could support the test cases where we differ the lengths of the various fields. Let me know if you had something else in mind!

piersy commented on op-chain-ops/cmd/celo-migrate/continuity.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+// CheckLengths makes sure the number of elements retrieved from each table is the same
+func (r *RLPBlockRange) CheckLengths(expectedLength uint64) error {
+	var err error
+	if uint64(len(r.hashes)) != expectedLength {
+		err = fmt.Errorf("Expected count mismatch in block range hashes: expected %d, actual %d", expectedLength, len(r.hashes))
+	}
+	if uint64(len(r.bodies)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range bodies: expected %d, actual %d", expectedLength, len(r.bodies)))
+	}
+	if uint64(len(r.headers)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range headers: expected %d, actual %d", expectedLength, len(r.headers)))
+	}
+	if uint64(len(r.receipts)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range receipts: expected %d, actual %d", expectedLength, len(r.receipts)))
+	}
+	if uint64(len(r.tds)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range total difficulties: expected %d, actual %d", expectedLength, len(r.tds)))
+	}
+	return err
+}
```
This error message is slightly confusing.

If read one way it can make it sound like  a "count mismatch" was expected. I'd suggest something like:

```
Unexpected length for block range.....
```

piersy commented:
Hey @alecps, I see what you're trying to do here, but this has become quite huge, 861 lines to check that each block follows it's parent.

The db continuity check runs pretty fast for me 1 minute for alfajores, so I would be in favour of reverting changes to the pre-existing migration code and just adding the db check as a separate command. This gives us the benefit of not having to modify the audited migration code. I don't want to be unilaterally deciding this though so please chime in @karlb & @palango

alecps commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -216,7 +216,7 @@ func writeAncientBlocks(ctx context.Context, freezer *rawdb.Freezer, in <-chan R
 			}
 			blockRangeEnd := blockRange.start + uint64(len(blockRange.hashes)) - 1
-			log.Info("Wrote ancient blocks", "start", blockRange.start, "end", blockRangeEnd, "count", len(blockRange.hashes), "remaining", totalAncientBlocks-blockRangeEnd)
+			log.Info("Wrote ancient blocks", "start", blockRange.start, "end", blockRangeEnd, "count", len(blockRange.hashes), "remaining", totalAncientBlocks-(blockRangeEnd+1))
```
This indexing error was missed by the audit

alecps commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -104,55 +94,65 @@ func migrateAncientsDb(ctx context.Context, oldDBPath, newDBPath string, batchSi
-				return fmt.Errorf("failed to read receipts from old freezer: %w", err)
-			}
-			blockRange.tds, err = freezer.AncientRange(rawdb.ChainFreezerDifficultyTable, start, count, 0)
+			blockRange, err := loadAncientRange(freezer, start, count)
```
@piersy This is the only refactor to the migration code path from the audited code that I haven't reverted. It's straightforward enough that I thought it might be okay to keep, since we need to use loadAncientRange in the continuity script. Happy to revert it though if you think that's best

alecps commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -104,55 +94,65 @@ func migrateAncientsDb(ctx context.Context, oldDBPath, newDBPath string, batchSi
 			return ctx.Err()
 		default:
-			count := min(batchSize, endBlock-i+1)
+			count := min(batchSize, endBlock-i)
```
The audit missed this indexing error. It should be `endBlock - i` because `endBlock` is not inclusive. This didn't cause any problems before because the `freezer.AncientRange` function returns at most `count` elements, and this indexing error could only cause a problem when reading the last section of the ancients (when the number of remaining blocks is less than `batchSize`)

alecps commented on op-chain-ops/cmd/celo-migrate/db.go:
```diff
@@ -61,7 +92,7 @@ func openDB(chaindataPath string, readOnly bool) (ethdb.Database, error) {
 // Opens a database without access to AncientsDb
 func openDBWithoutFreezer(chaindataPath string, readOnly bool) (ethdb.Database, error) {
-	if _, err := os.Stat(chaindataPath); errors.Is(err, os.ErrNotExist) {
+	if _, err := os.Stat(chaindataPath); err != nil {
```
Fixing this type of error check was part of the audit feedback. I fixed several others in [this PR](https://github.com/celo-org/optimism/pull/283/files), but somehow missed this one.

alecps commented:
@piersy Okay I've reverted all the changes to the migration code path. What remains is just the code for the continuity script, the unit tests for CheckContinuity, a fix for an indexing error, and some other super minor drive-by changes. 

I've moved the rest of the changes that weave the continuity checks into the migration logic to [alecps/addContinuityCheckToMigration](https://github.com/celo-org/optimism/pull/297)

alecps commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -104,56 +94,67 @@ func migrateAncientsDb(ctx context.Context, oldDBPath, newDBPath string, batchSi
-
 	for i := startBlock; i < endBlock; i += batchSize {
-		count := min(batchSize, endBlock-i+1)
+		count := min(batchSize, endBlock-i)
```
The audit missed this indexing error. It should be `endBlock - i` because endBlock is not inclusive. This didn't cause any problems before because the `freezer.AncientRange` function returns _at most_ `count` elements, and this indexing error could only cause a problem when reading the last section of the ancients (when the number of remaining blocks is less than `batchSize`)

alecps commented on op-chain-ops/cmd/celo-migrate/ancients.go:
```diff
@@ -104,56 +94,67 @@ func migrateAncientsDb(ctx context.Context, oldDBPath, newDBPath string, batchSi
-		var err error
-
-		blockRange.hashes, err = freezer.AncientRange(rawdb.ChainFreezerHashTable, start, count, 0)
+		blockRange, err := loadAncientRange(freezer, start, count)
```
@piersy This is the only refactor to the migration code path from the audited code that I haven't reverted. It's straightforward enough that I thought it might be okay to keep, since we need to use loadAncientRange in the continuity script. Happy to revert it though if you think that's best

palango commented on op-chain-ops/cmd/celo-migrate/continuity.go:
```diff
@@ -0,0 +1,127 @@
+// Follows checks if the current block has a number one greater than the previous block
+// and if the parent hash of the current block matches the hash of the previous block.
+func (e *RLPBlockElement) Follows(prev *RLPBlockElement) (err error) {
+	if e.Header().Number.Uint64() != prev.Header().Number.Uint64()+1 {
```
There's 12 instances of `.Header().Number.Uint64()`, maybe worth putting this in a small function for legibility

piersy commented:
Hey @alecps, I had a go at re-factoring this to make it a bit smaller. You can find the result [here](https://github.com/celo-org/optimism/pull/300). My feeling is it's a little simpler to follow the main continuity check flow, and a tiny bit smaller, but nothing particularly significant.

piersy commented on op-chain-ops/cmd/celo-migrate/main.go:
```diff
@@ -449,6 +495,138 @@ func runStateMigration(newDBPath string, opts stateMigrationOptions) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(8)
```
```suggestion
	// Use double the CPUs to account for the fact that each routine will block while reading from the db.
	g.SetLimit(runtime.NumCPU() * 2)
```

piersy reviewed (approved):
It looks like a lot of code for just checking the continuity, but in fact it is checking a lot of things and as we saw in (https://github.com/celo-org/optimism/pull/300) refactoring didn't manage to reduce the size significantly. So I think this is good to go!
